### PR TITLE
AvoidCopyAlways Check

### LINKED ISSUE #7654: New diagnostic proposal: erroring on CopyToOutputDirectory="Always"
One of the common pain points I've hit is builds re-running when they don't need to and a leading cause of this is `CopyToOutputDirectory="Always"` set somewhere on a Compile/None/EmbeddedResource/etc.

These are already collected up in `GetCopyToOutputDirectoryItems` so I propose we add a diagnostic (opt-in) to let people error on this in the base SDK (`Microsoft.CurrentVersion.targets`). Here's my rudamentary attempt at such a Task working locally:

```xml
<Target Name="ComplainAboutCopyAlways" 
        AfterTargets="_GetCopyToOutputDirectoryItemsFromThisProject"
        BeforeTargets="GetCopyToPublishDirectoryItems">
  <CallTarget Targets="_GetCopyToOutputDirectoryItemsFromThisProject">
    <Output TaskParameter="TargetOutputs" ItemName="_ThisProjectItemsToCopyToOutputDirectory" />
  </CallTarget>
  
  <!-- Note: due to this being an error, it only errors on the first...not sure if we can make it error on many correctly... -->
  <Error Text="Item '%(_ThisProjectItemsToCopyToOutputDirectory.TargetPath)' set as CopyToOutputDirectory=&quot;Always&quot;, use CopyToOutputDirectory=&quot;PreserveNewest&quot; instead." 
          Condition="'%(_ThisProjectItemsToCopyToOutputDirectory.CopyToOutputDirectory)'=='Always'" />
</Target>
```

This produces build output when violations are detected:
```
C:\path\Directory.Build.targets(382,5): error : Item 'AlwaysCopiedFile.Txt' set as CopyToOutputDirectory="Always", use CopyToOutputDirectory="PreserveNewest" instead. [C:\path\Test.csproj]
```

I think this would be very useful in the base targets, opted in via some new variable (adding a condition to my above example) - thoughts on this? I'd be happy to PR it if this is amenable we get a good variable and error message (I just tried to make an initial stab). Thanks!

/cc @baronfel @rainersigwald

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #7654

### Context
Adding a check validating that `Always` is not being used for `CopyToOutputDirectory` metadata. Or that the `SkipUnchangedFilesOnCopyAlways` property is set to `True` to avoid unnecessary copying.

### Testing
Tailored tests added - covering as well the #11052

### CODE REVIEW

File: src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        //  for this reason we use a hashset with already seen projects.
+        if (!_projectsSeen.Add(context.Data.ProjectFilePath))
+        {
+            return;
+        }
+
+        foreach (ItemData itemData in context.Data.EnumerateItemsOfTypes([ItemNames.Content, ItemNames.Compile, ItemNames.None, ItemNames.EmbeddedResource]))
+        {
+            // itemData.Type
+            // itemData.EvaluatedInclude
```
Comment (rainersigwald) [inline src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs:74]: ?

File: src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            foreach (KeyValuePair<string, string> keyValuePair in itemData.EnumerateMetadata())
+            {
+                if (MSBuildNameIgnoreCaseComparer.Default.Equals(keyValuePair.Key, ItemMetadataNames.copyToOutputDirectory))
+                {
+                    if (MSBuildNameIgnoreCaseComparer.Default.Equals(keyValuePair.Value, ItemMetadataNames.copyAlways))
+                    {
+                        // Project {0} specifies '{0}' item '{1}', ...
+                        context.ReportResult(BuildCheckResult.Create(
+                            SupportedRule,
+                            // Populating precise location tracked via https://github.com/orgs/dotnet/projects/373/views/1?pane=issue&itemId=58661732
```
Comment (rainersigwald) [inline src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs:85]: nit: can we link to an issue here? deeplinking a project feels bad to me.

File: src/Build/Resources/Strings.resx [hunk 3]
```diff
@@ -2176,11 +2176,19 @@ Utilization:          {0} Average Utilization: {1:###.0}</value>
...
-	<comment>Terms in quotes are not to be translated.</comment>
+	  <comment>Terms in quotes are not to be translated.</comment>
   </data>
   <data name="BuildCheck_BC0105_MessageFmt" xml:space="preserve">
     <value>Project {0} specifies 'EmbeddedResource' item '{1}', that has possibly a culture denoting extension ('{2}'), but explicit 'Culture' nor 'WithCulture=false' metadata are not specified.</value>
-	<comment>Terms in quotes are not to be translated.</comment>
+	  <comment>Terms in quotes are not to be translated.</comment>
+  </data>
+  <data name="BuildCheck_BC0106_Title" xml:space="preserve">
+    <value>It is recommended to avoid specifying 'Always' value of metadata 'CopyToOutputDirectory' as this can lead to unnecessary build performance degradation. Use 'PreserveNewest' or 'IfDifferent' metadata value, or set the 'SkipUnchangedFilesOnCopyAlways' property to true to employ more effective copying.</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:2186]: Please update with the final text since I proposed edits in the .md

**Review by maridematte — APPROVED** [review]

File: src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    internal override bool IsBuiltIn => true;
+
+    private readonly HashSet<string> _projectsSeen = new(MSBuildNameIgnoreCaseComparer.Default);
+
+    private void EvaluatedPropertiesAction(BuildCheckDataContext<EvaluatedPropertiesCheckData> context)
+    {
+        // We want to avoid repeated checking of a same project (as it might be evaluated multiple times)
+        //  for this reason we use a hashset with already seen projects.
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BuildCheck/Checks/CopyAlwaysCheck.cs:42]: I added explicit comment to bug to search all the checks - not to rely on comments than can get stale or be frogotten

**Review by rainersigwald — APPROVED** [review]

**Review by YuliiaKovalova — APPROVED** [review]
Looks good to me!

**Review by donJoseLuis — APPROVED** [review]

**Pull request merged**